### DISCUSSION TITLE
Serenity NavigationLink doesnt work after publish on IIS

### DISCUSSION THREAD

**Discussion by hafizTahir** [serenity-is/Serenity#5595] [Help]
after publishing on IIS its base address is not include in navigation that causes error:

Navigation.cs :
[assembly: NavigationMenu(int.MaxValue, "AMSDB", icon: "fa-laptop")]

[assembly: NavigationLink(int.MaxValue, "AMSDB/Category", typeof(MyPages.CategoryController), icon: null)]

![image](https://user-images.githubusercontent.com/39436753/112455072-2ff4d380-8d7b-11eb-9054-a43554c88f0f.png)

**Comment by wchitirobho** [comment]
comment this out

//Navigation.cs :
//[assembly: NavigationMenu(int.MaxValue, "AMSDB", icon: "fa-laptop")]
//[assembly: NavigationLink(int.MaxValue, "AMSDB/Category", typeof(MyPages.CategoryController), icon: null)]


 Create a new class and put this
use this to create dynamic url that includes your IIS Application name.

public class CategoryNavigation : INavigationItemSource
{
    public List<NavigationItemAttribute> GetItems()
    {
        var IISApplicationName = "AMSDB";
        var prefix = ("~/" + IISApplicationName + "/");

        var items = new List<NavigationItemAttribute>
            {
                new NavigationMenuAttribute(4000, "AMSDB", "icon-speedometer")
            };

        items.Add(new NavigationLinkAttribute(4000, path: "AMSDB/Category", url: prefix + "Default/Category/Index", permission: PermissionKeys.CategoryViewPermission, icon: ""));
        items.Add(new NavigationLinkAttribute(4000, path: "AMSDB/Sub Category", url: prefix + "Default/SubCategory/Index", permission: PermissionKeys.CategoryViewPermission, icon: ""));

        return items;
    }
}

**Reply by hafizTahir** [reply]
this method disturbs the navigation when i run it into iis express (debug mode of vs)

**Reply by wchitirobho** [reply]
it works, instead put your application name in the environment variable.

replace: 
var IISApplicationName = "AMSDB";
with:
var IISApplicationName  = Environment.GetEnvironmentVariable("iisappname");


Then in your main iis main application web config create location tag with the application name.

<location path="AMSDB">
	<system.webServer>
		<aspNetCore>
		  <environmentVariables>
			<environmentVariable name="iisappname" value="AMSDB" />
		  </environmentVariables>
		</aspNetCore>
	</system.webServer>
  </location>

then set an environment variable in your visual studio and set the environment value to nothing on your local dev, so that iis express does not look for the app name

**Reply by hafizTahir** [reply]
not working as expected other issues with this solution 
1- menu got collapsed back
2- i have to change for all modules(i.e Administration)

**Comment by VictorTomaili** [comment]
Configure your application base path on startup file
https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.usepathbaseextensions.usepathbase?view=aspnetcore-5.0

**Reply by hafizTahir** [reply]
please can explain with an example?

**Reply by VictorTomaili** [reply]
https://stackoverflow.com/questions/46593999/how-to-host-a-asp-net-core-application-under-a-sub-folder

**Reply by hafizTahir** [reply]
Its not getting apphost name(base name). i have tried both methods.
Need help

**Reply by hafizTahir** [reply]
> https://stackoverflow.com/questions/46593999/how-to-host-a-asp-net-core-application-under-a-sub-folder

did not get this. please can you help me resolving this issue

**Reply by hafizTahir** [reply]
@VictorTomaili

**Reply by VictorTomaili** [reply]
Try to start from there 
https://www.codeproject.com/Tips/1069787/Understanding-Startup-cs-file-in-ASP-NET
I can't help to write your code. I just navigate you to how to solve. I believe `UsePathBase` is the exact code that you need for navigation. Just research code samples on Google, StackOverflow, etc.

**Comment by amkh1900** [comment]
can you help me>>
how i can get current  **dispayName in respository** page
i try to use this code :
     var user2 = User.GetUserDefinition<UserDefinition>(UserRetriever);
            var omar = user2.DisplayName;
   

but it give error . i WORK SERENITY CORE
thanks .


i can get name of the current user by this
 var user = (UserDefinition)Authorization.UserDefinition;
AND var user = Context.User.Identity.Name;

**Reply by VictorTomaili** [reply]
Hey, @amkh1900 this is not a topic for your new questions. If you want to ask a new question open a new discussion and wait for a response. Don't spam on their topics. Be respectful to the community, please. Also, your question is already answered in other discussion topics. Just search you have a search text area on the top left of this page.

**Reply by hafizTahir** [reply]
@VictorTomaili can you guide me please!

**Reply by hafizTahir** [reply]
when i generate publish version for IIS using this :- 
**dotnet publish --output d:\SerenAMS -- configuration Release**
the base path is not included in url when i  click in navigation menu as showin in above picture attached

**Comment by Febriantos** [comment]
Go to NavigationModels.cs > NavigationModel function. 
Modify this code from "VirtualPathUtility.ToAbsolute(PathBase, x)" to "VirtualPathUtility.ToAbsolute(pathBase, x)".
Tested using Serenity 5.0.51.
